File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        
        if (start < 0 || start >= arr.size() || arr[start] < 0)
            return false;
        if (arr[start] == 0)
            return true;
        
        int dist = arr[start];
        arr[start] = -1;
        
        if (canReach(arr, start - dist) || canReach(arr, start + dist))
            return true;
        return false;
    }
};

File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        if start < 0 or start >= len(arr) or arr[start] < 0:
            return False
        if arr[start] == 0:
            return True
        
        dist = arr[start]
        arr[start] = -1
        
        if self.canReach(arr, start - dist) or self.canReach(arr, start + dist):
            return True
        return False
        